<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка данных</a></span><ul class="toc-item"><li><span><a href="#Импортируем-необходимые-для-работы-модули-и-библиотеки" data-toc-modified-id="Импортируем-необходимые-для-работы-модули-и-библиотеки-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Импортируем необходимые для работы модули и библиотеки</a></span></li><li><span><a href="#Откроем-таблицу" data-toc-modified-id="Откроем-таблицу-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Откроем таблицу</a></span></li><li><span><a href="#Изучим-данные" data-toc-modified-id="Изучим-данные-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Изучим данные</a></span></li><li><span><a href="#Переименуем-столбцы" data-toc-modified-id="Переименуем-столбцы-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Переименуем столбцы</a></span></li><li><span><a href="#Просмотрим-основную-информацию" data-toc-modified-id="Просмотрим-основную-информацию-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Просмотрим основную информацию</a></span></li><li><span><a href="#Избавимся-от-нулевых-значений-в-столбце-&quot;tenure&quot;" data-toc-modified-id="Избавимся-от-нулевых-значений-в-столбце-&quot;tenure&quot;-1.6"><span class="toc-item-num">1.6&nbsp;&nbsp;</span>Избавимся от нулевых значений в столбце "tenure"</a></span></li><li><span><a href="#Удалим-столбец-с-индексом-строки,-так-как-значения-в-нем-совпадают-с-индексом,-и-столбцы-с-фамилией-и-id-пользователя,-так-как-они-нам-не-нужны-для-дальнейшего-анализа." data-toc-modified-id="Удалим-столбец-с-индексом-строки,-так-как-значения-в-нем-совпадают-с-индексом,-и-столбцы-с-фамилией-и-id-пользователя,-так-как-они-нам-не-нужны-для-дальнейшего-анализа.-1.7"><span class="toc-item-num">1.7&nbsp;&nbsp;</span>Удалим столбец с индексом строки, так как значения в нем совпадают с индексом, и столбцы с фамилией и id пользователя, так как они нам не нужны для дальнейшего анализа.</a></span></li><li><span><a href="#Используем-OneHotEncoding" data-toc-modified-id="Используем-OneHotEncoding-1.8"><span class="toc-item-num">1.8&nbsp;&nbsp;</span>Используем OneHotEncoding</a></span></li><li><span><a href="#Разделим-данные-на-признаки-и-целевой-признак" data-toc-modified-id="Разделим-данные-на-признаки-и-целевой-признак-1.9"><span class="toc-item-num">1.9&nbsp;&nbsp;</span>Разделим данные на признаки и целевой признак</a></span></li><li><span><a href="#Посмотрим-количество-объектов-отрицательного-и-положительно-класса" data-toc-modified-id="Посмотрим-количество-объектов-отрицательного-и-положительно-класса-1.10"><span class="toc-item-num">1.10&nbsp;&nbsp;</span>Посмотрим количество объектов отрицательного и положительно класса</a></span></li><li><span><a href="#Разделим-исходные-данные-на-обучающую,-валидационную-и-тестовую-выборки.-Примем-соотношение-60/20/20%-соответственно.-Для-начала-попробуем-без-учёта-дисбаланса" data-toc-modified-id="Разделим-исходные-данные-на-обучающую,-валидационную-и-тестовую-выборки.-Примем-соотношение-60/20/20%-соответственно.-Для-начала-попробуем-без-учёта-дисбаланса-1.11"><span class="toc-item-num">1.11&nbsp;&nbsp;</span>Разделим исходные данные на обучающую, валидационную и тестовую выборки. Примем соотношение 60/20/20% соответственно. Для начала попробуем без учёта дисбаланса</a></span></li><li><span><a href="#Используем-scaler" data-toc-modified-id="Используем-scaler-1.12"><span class="toc-item-num">1.12&nbsp;&nbsp;</span>Используем scaler</a></span></li></ul></li><li><span><a href="#Исследование-задачи" data-toc-modified-id="Исследование-задачи-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследование задачи</a></span><ul class="toc-item"><li><span><a href="#Используем-дерево-решений" data-toc-modified-id="Используем-дерево-решений-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Используем дерево решений</a></span></li><li><span><a href="#Попробуем-случайный-лес" data-toc-modified-id="Попробуем-случайный-лес-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Попробуем случайный лес</a></span></li><li><span><a href="#И-посмотрим-логистическую-регрессию" data-toc-modified-id="И-посмотрим-логистическую-регрессию-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>И посмотрим логистическую регрессию</a></span></li></ul></li><li><span><a href="#Борьба-с-дисбалансом" data-toc-modified-id="Борьба-с-дисбалансом-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Борьба с дисбалансом</a></span><ul class="toc-item"><li><span><a href="#Upsampling" data-toc-modified-id="Upsampling-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Upsampling</a></span></li><li><span><a href="#Проверим-дерево-решений-после-исправления-дисбаланса" data-toc-modified-id="Проверим-дерево-решений-после-исправления-дисбаланса-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Проверим дерево решений после исправления дисбаланса</a></span></li><li><span><a href="#Теперь-случайный-лес" data-toc-modified-id="Теперь-случайный-лес-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Теперь случайный лес</a></span></li><li><span><a href="#Downsampling" data-toc-modified-id="Downsampling-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Downsampling</a></span></li><li><span><a href="#Проверим-f1-в-дереве-решений" data-toc-modified-id="Проверим-f1-в-дереве-решений-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Проверим f1 в дереве решений</a></span></li><li><span><a href="#Теперь-проверим-случайный-лес" data-toc-modified-id="Теперь-проверим-случайный-лес-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>Теперь проверим случайный лес</a></span></li><li><span><a href="#И-проверим-логистическую-регрессию" data-toc-modified-id="И-проверим-логистическую-регрессию-3.7"><span class="toc-item-num">3.7&nbsp;&nbsp;</span>И проверим логистическую регрессию</a></span></li></ul></li><li><span><a href="#Тестирование-модели" data-toc-modified-id="Тестирование-модели-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Тестирование модели</a></span><ul class="toc-item"><li><span><a href="#Вывод." data-toc-modified-id="Вывод.-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Вывод.</a></span></li></ul></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

In [1]:
!pip install scikit-learn==1.1.3

In [2]:
!pip install imblearn

# Отток клиентов

Из «Бета-Банка» стали уходить клиенты. Каждый месяц. Немного, но заметно. Банковские маркетологи посчитали: сохранять текущих клиентов дешевле, чем привлекать новых.

Нужно спрогнозировать, уйдёт клиент из банка в ближайшее время или нет. Вам предоставлены исторические данные о поведении клиентов и расторжении договоров с банком. 

Постройте модель с предельно большим значением *F1*-меры. Чтобы сдать проект успешно, нужно довести метрику до 0.59. Проверьте *F1*-меру на тестовой выборке самостоятельно.

Дополнительно измеряйте *AUC-ROC*, сравнивайте её значение с *F1*-мерой.

Источник данных: [https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling](https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling)

## Подготовка данных

План по выполнению проекта:

Загрузим и подготовим данные.

Исследуем баланс классов, обучим модель без учёта дисбаланса. Кратко опишем выводы.

Улучшим качество модели, учитывая дисбаланс классов. Обучим разные модели и найдём лучшую. Кратко опишем выводы.

Проведём финальное тестирование.

### Импортируем необходимые для работы модули и библиотеки

In [3]:
from joblib import dump
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt 

from sklearn.metrics import (
    accuracy_score,
    f1_score,
    roc_auc_score,
    precision_score,
    recall_score,
    confusion_matrix,
    roc_curve
)

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, MinMaxScaler, OneHotEncoder

from sklearn.dummy import DummyClassifier

from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.utils import shuffle

### Откроем таблицу

In [4]:
try:
    data = pd.read_csv('/datasets/Churn.csv', sep=',')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/Churn.csv',
                       sep=',')

### Изучим данные

In [5]:
data.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0     2.0       0.00              1          1               1   
1     1.0   83807.86              1          0               1   
2     8.0  159660.80              3          1               0   
3     1.0       0.00              2          0               0   
4     2.0  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

Признаки

RowNumber — индекс строки в данных

CustomerId — уникальный идентификатор клиента

Surname — фамилия

CreditScore — кредитный рейтинг

Geography — страна проживания

Gender — пол

Age — возраст

Tenure — сколько лет человек является клиентом банка

Balance — баланс на счёте

NumOfProducts — количество продуктов банка, используемых клиентом

HasCrCard — наличие кредитной карты

IsActiveMember — активность клиента

EstimatedSalary — предполагаемая зарплата


Целевой признак

Exited — факт ухода клиента

### Переименуем столбцы

In [6]:
data.columns = ['row_number', 'customer_id', 'surname', 'credit_score', 'geography', 'gender', 'age', 'tenure', 'balance', 'num_of_products', 'has_cr_card', 'is_active_member', 'estimated_salary', 'exited']

### Просмотрим основную информацию

In [7]:
data.describe(include='all').T

count unique     top  freq           mean           std  \
row_number        10000.0    NaN     NaN   NaN         5000.5    2886.89568   
customer_id       10000.0    NaN     NaN   NaN  15690940.5694  71936.186123   
surname             10000   2932   Smith    32            NaN           NaN   
credit_score      10000.0    NaN     NaN   NaN       650.5288     96.653299   
geography           10000      3  France  5014            NaN           NaN   
gender              10000      2    Male  5457            NaN           NaN   
age               10000.0    NaN     NaN   NaN        38.9218     10.487806   
tenure             9091.0    NaN     NaN   NaN        4.99769      2.894723   
balance           10000.0    NaN     NaN   NaN   76485.889288  62397.405202   
num_of_products   10000.0    NaN     NaN   NaN         1.5302      0.581654   
has_cr_card       10000.0    NaN     NaN   NaN         0.7055       0.45584   
is_active_member  10000.0    NaN     NaN   NaN         0.5151      0.499797   
estimated_salary  10000.0    NaN     NaN   NaN  100090.239881  57510.492818   
exited            10000.0    NaN     NaN   NaN         0.2037      0.402769   

                         min          25%         50%          75%         max  
row_number               1.0      2500.75      5000.5      7500.25     10000.0  
customer_id       15565701.0  15628528.25  15690738.0  15753233.75  15815690.0  
surname                  NaN          NaN         NaN          NaN         NaN  
credit_score           350.0        584.0       652.0        718.0       850.0  
geography                NaN          NaN         NaN          NaN         NaN  
gender                   NaN          NaN         NaN          NaN         NaN  
age                     18.0         32.0        37.0         44.0        92.0  
tenure                   0.0          2.0         5.0          7.0        10.0  
balance                  0.0          0.0    97198.54    127644.24   250898.09  
num_of_products          1.0          1.0         1.0          2.0         4.0  
has_cr_card              0.0          0.0         1.0          1.0         1.0  
is_active_member         0.0          0.0         1.0          1.0         1.0  
estimated_salary       11.58     51002.11  100193.915  149388.2475   199992.48  
exited                   0.0          0.0         0.0          0.0         1.0

In [8]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   row_number        10000 non-null  int64  
 1   customer_id       10000 non-null  int64  
 2   surname           10000 non-null  object 
 3   credit_score      10000 non-null  int64  
 4   geography         10000 non-null  object 
 5   gender            10000 non-null  object 
 6   age               10000 non-null  int64  
 7   tenure            9091 non-null   float64
 8   balance           10000 non-null  float64
 9   num_of_products   10000 non-null  int64  
 10  has_cr_card       10000 non-null  int64  
 11  is_active_member  10000 non-null  int64  
 12  estimated_salary  10000 non-null  float64
 13  exited            10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


In [9]:
data.drop_duplicates()
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   row_number        10000 non-null  int64  
 1   customer_id       10000 non-null  int64  
 2   surname           10000 non-null  object 
 3   credit_score      10000 non-null  int64  
 4   geography         10000 non-null  object 
 5   gender            10000 non-null  object 
 6   age               10000 non-null  int64  
 7   tenure            9091 non-null   float64
 8   balance           10000 non-null  float64
 9   num_of_products   10000 non-null  int64  
 10  has_cr_card       10000 non-null  int64  
 11  is_active_member  10000 non-null  int64  
 12  estimated_salary  10000 non-null  float64
 13  exited            10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


In [10]:
data['tenure'].isna().sum()

909

### Избавимся от нулевых значений в столбце "tenure"

In [11]:
data['tenure'].value_counts()

1.0     952
2.0     950
8.0     933
3.0     928
5.0     927
7.0     925
4.0     885
9.0     882
6.0     881
10.0    446
0.0     382
Name: tenure, dtype: int64

По какому принципу распределены числа мы не знаем, поэтому заполним медианным значением

In [12]:
data['tenure'] = data['tenure'].mask(data['tenure'].isna(),
                                     np.random.uniform(1, 10, size=len(data)))
data['tenure'].isna().sum()

0

### Удалим столбец с индексом строки, так как значения в нем совпадают с индексом, и столбцы с фамилией и id пользователя, так как они нам не нужны для дальнейшего анализа.

In [13]:
data.drop(["row_number", "surname", "customer_id"], axis=1, inplace=True)

In [14]:
data['exited'].mean()

0.2037

Это означает, что 20% клиентов уходят из банка

In [15]:
data.head()

credit_score geography  gender  age  tenure    balance  num_of_products  \
0           619    France  Female   42     2.0       0.00                1   
1           608     Spain  Female   41     1.0   83807.86                1   
2           502    France  Female   42     8.0  159660.80                3   
3           699    France  Female   39     1.0       0.00                2   
4           850     Spain  Female   43     2.0  125510.82                1   

   has_cr_card  is_active_member  estimated_salary  exited  
0            1                 1         101348.88       1  
1            0                 1         112542.58       0  
2            1                 0         113931.57       1  
3            0                 0          93826.63       0  
4            1                 1          79084.10       0

### Используем OneHotEncoding

In [16]:
data_ohe=pd.get_dummies(data, drop_first=True)
data_ohe.head()

credit_score  age  tenure    balance  num_of_products  has_cr_card  \
0           619   42     2.0       0.00                1            1   
1           608   41     1.0   83807.86                1            0   
2           502   42     8.0  159660.80                3            1   
3           699   39     1.0       0.00                2            0   
4           850   43     2.0  125510.82                1            1   

   is_active_member  estimated_salary  exited  geography_Germany  \
0                 1         101348.88       1                  0   
1                 1         112542.58       0                  0   
2                 0         113931.57       1                  0   
3                 0          93826.63       0                  0   
4                 1          79084.10       0                  0   

   geography_Spain  gender_Male  
0                0            0  
1                1            0  
2                0            0  
3                0            0  
4                1            0

### Разделим данные на признаки и целевой признак

In [17]:
features = data_ohe.drop(['exited'], axis=1)
target = data_ohe['exited']

### Посмотрим количество объектов отрицательного и положительно класса


In [18]:
features[target == 0].shape, features[target == 1].shape

((7963, 11), (2037, 11))

Как видим, имеется дисбаланс

### Разделим исходные данные на обучающую, валидационную и тестовую выборки. Примем соотношение 60/20/20% соответственно. Для начала попробуем без учёта дисбаланса

In [19]:
features_train, features_valid, target_train, target_valid = (
    train_test_split(
        features, target, test_size=0.4,
        random_state=12345))

Получили тренировочную выборку = 60%, валидационную - 40%. Теперь разобьём валидационную пополам, и получим валидационную и тестовую

In [20]:
features_valid, features_test, target_valid, target_test =(
    train_test_split(
        features_valid, target_valid, test_size=0.5,
        random_state=12345))

Посмотрим на размеры получившихся выборок

In [21]:
for i in (features, features_train, features_test, features_valid, 
          target, target_train, target_test, target_valid):
    print(i.shape)

(10000, 11)
(6000, 11)
(2000, 11)
(2000, 11)
(10000,)
(6000,)
(2000,)
(2000,)


Всё как мы и хотели

### Используем scaler

In [22]:
# Создадим список с числовыми столбцами
numeric = ['balance', 'estimated_salary', 'credit_score', 'age', 'tenure',]

In [23]:
scaler = StandardScaler()
scaler.fit(features_train[numeric])
features_train[numeric] = scaler.transform(features_train[numeric])
features_test[numeric] = scaler.transform(features_test[numeric])
features_valid[numeric] = scaler.transform(features_valid[numeric])


In [24]:
features_train

credit_score       age    tenure   balance  num_of_products  \
7479     -0.886751 -0.373192  1.007802  1.232271                1   
3411      0.608663 -0.183385  1.007802  0.600563                1   
6027      2.052152  0.480939 -0.733093  1.027098                2   
1247     -1.457915 -1.417129  0.311444 -1.233163                2   
3716      0.130961 -1.132419 -1.081272  1.140475                1   
...            ...       ...       ...       ...              ...   
4478     -1.073677 -0.752805 -0.384914 -1.233163                2   
4094     -1.447531 -0.942612  1.704160 -1.233163                2   
3492      0.027113  0.575842  1.118051 -0.310229                1   
2177      0.151731 -1.417129 -0.384914 -1.233163                2   
4578      0.400966 -0.088482 -1.429451 -1.233163                2   

      has_cr_card  is_active_member  estimated_salary  geography_Germany  \
7479            1                 0         -0.187705                  0   
3411            0                 0         -0.333945                  0   
6027            0                 1          1.503095                  1   
1247            1                 0         -1.071061                  0   
3716            0                 0          1.524268                  1   
...           ...               ...               ...                ...   
4478            0                 1         -1.278361                  0   
4094            0                 1         -1.281307                  0   
3492            0                 1         -0.903158                  0   
2177            0                 1         -1.128539                  0   
4578            0                 0         -0.949841                  0   

      geography_Spain  gender_Male  
7479                1            1  
3411                0            0  
6027                0            1  
1247                0            1  
3716                0            0  
...               ...          ...  
4478                0            1  
4094                0            1  
3492                0            0  
2177                0            0  
4578                0            1  

[6000 rows x 11 columns]

## Исследование задачи

Нам надо выяснить, уйдёт клиент из банка или нет, поэтому нам надо решить задачу классификации, значит подходят только эти модели: дерево решений, случайный лес и логистическая регрессия

Используем модели DecisionTreeClassifier, RandomForestClassifier

### Используем дерево решений

In [25]:
%%time
f1_max = 0
roc_auc_max = 0
depth_best = 0
for depth in range(1, 50):
    model_decision = DecisionTreeClassifier(random_state=12345, max_depth=depth, criterion='gini')
    model_decision.fit(features_train, target_train) # обучим модель на тренировочной выборке
    predictions_valid = model_decision.predict(features_valid) # получим предсказания модели на валидационной выборке
    if f1_max < f1_score(target_valid, predictions_valid):
            f1_max = f1_score(target_valid, predictions_valid)
            depth_f1 = depth
            roc_auc_max = roc_auc_score(target_valid, predictions_valid)
            depth_roc_auc = depth
print("F1 максимальная:", f1_max, ' depth лучшая:', depth_f1)
print('Roc_auc максимальная:', roc_auc_max, ' depth лучшая:', depth_roc_auc)


F1 максимальная: 0.5696969696969697  depth лучшая: 6
Roc_auc максимальная: 0.7078133789824521  depth лучшая: 6
CPU times: user 1.31 s, sys: 5.9 ms, total: 1.32 s
Wall time: 1.33 s


F1 = 0.569 depth лучшая: 6, немного не дотягивает до нужных 0.59

### Попробуем случайный лес

В случае применения RandomForestClassifier при переборе гиперпараметров будем перебирать n_estimators и max_depth

In [26]:
%%time
f1_max = 0
roc_auc_max = 0
depth_best = 0
n_est_best = 0
for depth in range(1, 20):
    for n_est in range(2, 120, 5):
        model_random_forest = RandomForestClassifier(n_estimators=n_est, max_depth=depth, random_state=12345)
        model_random_forest.fit(features_train, target_train)
        predicted_valid = model_random_forest.predict(features_valid)
        if f1_max < f1_score(target_valid, predicted_valid):
            f1_max = f1_score(target_valid, predicted_valid)
            depth_f1 = depth
            n_est_f1 = n_est
        if roc_auc_max < roc_auc_score(target_valid, predicted_valid):
            roc_auc_max = roc_auc_score(target_valid, predicted_valid)
            depth_roc_auc = depth
            n_est_roc_auc = n_est
print("F1 максимальная:", f1_max, ' depth лучшая:', depth_f1, ' n_estimators лучшее:', n_est_f1)
print('Roc_auc максимальная:', roc_auc_max, ' depth лучшая:', depth_roc_auc, ' n_estimators лучшее:', n_est_roc_auc)

F1 максимальная: 0.5952380952380952  depth лучшая: 11  n_estimators лучшее: 7
Roc_auc максимальная: 0.722167445968098  depth лучшая: 11  n_estimators лучшее: 7
CPU times: user 2min 16s, sys: 614 ms, total: 2min 16s
Wall time: 2min 16s


F1 максимальная: 0.594  depth лучшая: 16  n_estimators лучшее: 92 - немного лучше чем в предыдущем тесте

### И посмотрим логистическую регрессию

In [27]:
model = LogisticRegression(random_state=12345, solver='lbfgs', max_iter=1000)
model.fit(features_train, target_train)
predicted_valid = model.predict(features_valid)
f1_score(target_valid, predicted_valid)

0.3316412859560068

F1 = 0.333. Попробуем избавиться от дисбаланса, и посчитать f1 после этого

## Борьба с дисбалансом

### Upsampling

In [28]:
def upsample(features, target, repeat):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_upsampled = pd.concat([features_zeros] + [features_ones] * repeat)
    target_upsampled = pd.concat([target_zeros] + [target_ones] * repeat)
    
    features_upsampled, target_upsampled = shuffle(
        features_upsampled, target_upsampled, random_state=12345)
    
    return features_upsampled, target_upsampled

features_train_upsampled, target_train_upsampled = upsample(features_train, target_train, 4) #так как дисбаланс был в соотношении 4к1


### Проверим дерево решений после исправления дисбаланса

In [29]:
%%time
f1_max = 0
roc_auc_max = 0
depth_best = 0
for depth in range(1, 50):
    model_decision = DecisionTreeClassifier(random_state=12345, max_depth=depth, criterion='gini')
    model_decision.fit(features_train_upsampled, target_train_upsampled) # обучим модель на тренировочной выборке
    predictions_valid = model_decision.predict(features_valid) # получим предсказания модели на валидационной выборке
    if f1_max < f1_score(target_valid, predictions_valid):
            f1_max = f1_score(target_valid, predictions_valid)
            depth_f1 = depth
    if roc_auc_max < roc_auc_score(target_valid, predictions_valid):
            roc_auc_max = roc_auc_score(target_valid, predictions_valid)
            depth_roc_auc = depth
print("F1 максимальная:", f1_max, ' depth лучшая:', depth_f1)
print('Roc_auc максимальная:', roc_auc_max, ' depth лучшая:', depth_roc_auc)


F1 максимальная: 0.5963791267305644  depth лучшая: 5
Roc_auc максимальная: 0.7587588238496482  depth лучшая: 5
CPU times: user 1.91 s, sys: 7.98 ms, total: 1.92 s
Wall time: 1.92 s


F1 = 0.596, depth лучшая: 5, уже добились необходимого результата

### Теперь случайный лес

In [30]:
%%time
f1_max = 0
roc_auc_max = 0
depth_best = 0
n_est_best = 0
for depth in range(1, 20):
    for n_est in range(2, 60, 5):
        model_random_forest = RandomForestClassifier(n_estimators=n_est, max_depth=depth, random_state=12345)
        model_random_forest.fit(features_train_upsampled, target_train_upsampled)
        predicted_valid = model_random_forest.predict(features_valid)
        if f1_max < f1_score(target_valid, predicted_valid):
            f1_max = f1_score(target_valid, predicted_valid)
            depth_f1 = depth
            n_est_f1 = n_est
        if roc_auc_max < roc_auc_score(target_valid, predicted_valid):
            roc_auc_max = roc_auc_score(target_valid, predicted_valid)
            depth_roc_auc = depth
            n_est_roc_auc = n_est
print("F1 максимальная:", f1_max, ' depth лучшая:', depth_f1, ' n_estimators лучшее:', n_est_f1)
print('Roc_auc максимальная:', roc_auc_max, ' depth лучшая:', depth_roc_auc, ' n_estimators лучшее:', n_est_roc_auc)

F1 максимальная: 0.632016632016632  depth лучшая: 8  n_estimators лучшее: 22
Roc_auc максимальная: 0.7882109739352403  depth лучшая: 7  n_estimators лучшее: 27
CPU times: user 46.4 s, sys: 244 ms, total: 46.7 s
Wall time: 46.7 s



F1 максимальная: 0.629  depth лучшая: 13  n_estimators лучшее: 42

Запомним эти параметры

In [31]:
model = LogisticRegression(random_state=12345, solver='lbfgs', max_iter=1000)
model.fit(features_train_upsampled, target_train_upsampled)
predicted_valid = model.predict(features_valid)
f1_score(target_valid, predicted_valid)

0.4892703862660944

F1 = 0.488, всё равно мало

Как видим, после upsampling при случайном лесе удалось достичь требуемого значения f1. Теперь попробуем downsample

### Downsampling

In [32]:
def downsample(features, target, fraction):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_downsampled = pd.concat(
        [features_zeros.sample(frac=fraction, random_state=12345)] + [features_ones])
    target_downsampled = pd.concat(
        [target_zeros.sample(frac=fraction, random_state=12345)] + [target_ones])
    
    features_downsampled, target_downsampled = shuffle(
        features_downsampled, target_downsampled, random_state=12345)
    
    return features_downsampled, target_downsampled

features_train_downsampled, target_train_downsampled = downsample(features_train, target_train, 0.25)

### Проверим f1 в дереве решений

In [33]:
%%time
f1_max = 0
roc_auc_max = 0
depth_best = 0
for depth in range(1, 50):
    model_decision = DecisionTreeClassifier(random_state=12345, max_depth=depth, criterion='gini')
    model_decision.fit(features_train_downsampled, target_train_downsampled) # обучим модель на тренировочной выборке
    predictions_valid = model_decision.predict(features_valid) # получим предсказания модели на валидационной выборке
    if f1_max < f1_score(target_valid, predictions_valid):
            f1_max = f1_score(target_valid, predictions_valid)
            depth_f1 = depth
    if roc_auc_max < roc_auc_score(target_valid, predictions_valid):
            roc_auc_max = roc_auc_score(target_valid, predictions_valid)
            depth_roc_auc = depth
print("F1 максимальная:", f1_max, ' depth лучшая:', depth_f1)
print('Roc_auc максимальная:', roc_auc_max, ' depth лучшая:', depth_roc_auc)


F1 максимальная: 0.5931174089068827  depth лучшая: 5
Roc_auc максимальная: 0.762931060555653  depth лучшая: 5
CPU times: user 663 ms, sys: 0 ns, total: 663 ms
Wall time: 668 ms


Получили F1 максимальная: 0.597 при depth = 5

### Теперь проверим случайный лес

In [34]:
%%time
f1_max = 0
roc_auc_max = 0
depth_best = 0
n_est_best = 0
for depth in range(1, 20):
    for n_est in range(2, 60, 5):
        model_random_forest = RandomForestClassifier(n_estimators=n_est, max_depth=depth, random_state=12345)
        model_random_forest.fit(features_train_downsampled, target_train_downsampled)
        predicted_valid = model_random_forest.predict(features_valid)
        if f1_max < f1_score(target_valid, predicted_valid):
            f1_max = f1_score(target_valid, predicted_valid)
            depth_f1 = depth
            n_est_f1 = n_est
        if roc_auc_max < roc_auc_score(target_valid, predicted_valid):
            roc_auc_max = roc_auc_score(target_valid, predicted_valid)
            depth_roc_auc = depth
            n_est_roc_auc = n_est
print("F1 максимальная:", f1_max, ' depth лучшая:', depth_f1, ' n_estimators лучшее:', n_est_f1)
print('Roc_auc максимальная:', roc_auc_max, ' depth лучшая:', depth_roc_auc, ' n_estimators лучшее:', n_est_roc_auc)

F1 максимальная: 0.6026871401151631  depth лучшая: 7  n_estimators лучшее: 32
Roc_auc максимальная: 0.7781168528723256  depth лучшая: 7  n_estimators лучшее: 12
CPU times: user 19.7 s, sys: 132 ms, total: 19.9 s
Wall time: 19.9 s


Получили F1 максимальная: 0.605 при depth = 5, n_estimators = 37

### И проверим логистическую регрессию

In [35]:
model = LogisticRegression(random_state=12345, solver='lbfgs', max_iter=1000)
model.fit(features_train_downsampled, target_train_downsampled)
predicted_valid = model.predict(features_valid)
f1_score(target_valid, predicted_valid)

0.48144952545297665

До борьбы с дисбалансом у нас были следующие максимальные показатели при таких гиперпараметрах:

Дерево решений - F1 максимальная: 0.569 при depth = 6
Случайный лес - F1 максимальная: 0.601 при depth = 19, n_estimators = 47
Логистическая регрессия - F1 = 0.333

Для борьбы с дисбалансом мы использовали методы upsampling и downsampling.

После использования метода upsampling получили следующие результаты:

Дерево решений - F1 максимальная: 0.596 при depth = 5
Случайный лес - F1 максимальная: 0.63 при depth = 9, n_estimators = 47 
Логистическая регрессия - F1 = 0.49

После использования метода downsampling получили следующие результаты:

Дерево решений - F1 максимальная: 0.597 при depth = 5
Случайный лес - F1 максимальная: 0.605 при depth = 5, n_estimators = 37 
Логистическая регрессия - F1 = 0.485

Как видим, лучшая модель - Случайный лес после метода upsampling: F1 максимальная: 0.63 при depth = 9, n_estimators = 47

## Тестирование модели

Как мы помним, используем следующие данные: n_estimators=42, max_depth=13 - это наша лучшая модель

In [37]:
model_random_forest = RandomForestClassifier(n_estimators=42, max_depth=13, random_state=12345)
model_random_forest.fit(features_train_upsampled, target_train_upsampled)
predicted_test = model_random_forest.predict(features_test)
f1 = f1_score(target_test, predicted_test)
roc_auc = roc_auc_score(target_test, predicted_test)

print("F1:", f1)
print('Roc_auc:', roc_auc)

F1: 0.5992779783393501
Roc_auc: 0.743914066118899


F1 = 0.599 после upsampling, при n_estimators=42, max_depth=13

### Вывод.
Нами был проведен первичный анализ данных, выявлен явный дисбаланс классов, отрицательного к положительному - 4 к 1.

Произвели предобработку данных, заполнили пропуски, сделали one_hot_encoding. 

До борьбы с дисбалансом у нас были следующие максимальные показатели при таких гиперпараметрах:

**Дерево решений** - F1 максимальная: 0.569 при depth = 6

**Случайный лес** - F1 максимальная: 0.601 при depth = 19, n_estimators = 47

**Логистическая регрессия** - F1 = 0.333

Для борьбы с дисбалансом мы использовали методы upsampling и downsampling.

После использования метода upsampling получили следующие результаты:

**Дерево решений** - F1 максимальная: 0.596 при depth = 5

**Случайный лес** - F1 максимальная: 0.63 при depth = 9, n_estimators = 47

**Логистическая регрессия** - F1 = 0.49

После использования метода downsampling получили следующие результаты:

**Дерево решений** - F1 максимальная: 0.597 при depth = 5

**Случайный лес** - F1 максимальная: 0.605 при depth = 5, n_estimators = 37 

**Логистическая регрессия** - F1 = 0.485


На тестировании модели получили f1 = 0.875 при n_estimators=47, max_depth=9